File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

int solution(int x, int y, int n) {
    int answer = -1;
    
    vector<int> visited(1000001, -1);
    visited[x] = 0;
    queue<int> q;
    q.push(x);
    
    while(!q.empty()){
        int num = q.front();
        q.pop();
        
        if(num == y){
            answer = visited[y];
            break;
        }
        
        if(num*2 <= y && visited[num * 2] == -1){
            visited[num*2] = visited[num]+1;
            q.push(num*2);
        }
        if(num*3 <= y && visited[num * 3] == -1){
            visited[num*3] = visited[num]+1;
            q.push(num*3);
        }
        if(num+n <= y && visited[num + n] == -1){
            visited[num + n] = visited[num]+1;
            q.push(num + n);
        }
    }
    
    return answer;
}

File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.py
from collections import deque

def solution(x, y, n):
    answer = -1
    
    q = deque([x])
    visited = [-1] * (y+1)
    visited[x] = 0
    
    while q:
        num = q.popleft()
        
        if num == y:
            answer = visited[num]
            break
        
        if num * 2 <= y and visited[num*2] == -1:
            visited[num*2] = visited[num] + 1
            q.append(num*2)
        if num * 3 <= y and visited[num*3] == -1:
            visited[num*3] = visited[num] + 1
            q.append(num*3)
        if num + n <= y and visited[num + n] == -1:
            visited[num + n] = visited[num] + 1
            q.append(num + n)
    
    return answer